src/analyses/base.ml

```diff
@@ -1349,22 +1393,18 @@ struct
 
   let invariant ctx a gs st exp tv =
     let open Deriving.Cil in
-    let fallback reason =
```

Comment:
The way the fallback was called was incorrect, see e.g. `01/02 branch`:

```
  switch (x) {
    case 5: k = 3 + x; assert (x == 5); break; // reported as dead when using the fallback
    case 6: k = 2 + x; assert (x == 6); break; // reported as dead when using the fallback
    default: k = 8; assert(x != 5); assert( x!= 6);
  }
```
In think the issue there was that `fallback` would be called from within a recursive call to `inv_exp` and then used to constrain the wrong variable (e.g. in the `Binop` case), as fallback always takes the `exp` and `tv` from `invariant`.

Rather than fixing this, I thought it was nicer to have a clean separation were the different implementations of `invariant` are to distinct things and not not interact with each other. But that is something that is opened for debate I guess.

Reply:
Ok, yes, that sounds like a bug. No strong opinion - being able to dynamically fallback seemed more flexible/precise, but that only matters if we don't get rid of the old implementation. Should check which cases are not covered yet.

Reply:
I went through it again this morning, the issue I described above was something unrelated. So the fallback is not incorrect.

What confused me there is that we always call `fallback` for the original `exp` and `tv` that `invariant` was called with, and not with the `exp` in `inv_exp` that we failed with.
This is however not a problem, as fallback returns a new `cpa` (improved by the old `invariant` for the original `exp` and `tv`) that we then either meet with `inv_exp` for the other argument to a `binop` or just return.

I'll revert those changes.